sdks/python/apache_beam/runners/interactive/augmented_pipeline.py

```diff
@@ -0,0 +1,137 @@
+from apache_beam.runners.interactive.caching.write_cache import WriteCache
+
+
+class AugmentedPipeline:
+  """A pipeline with augmented interactive flavor that caches intermediate
+  PCollections defined by the user, reads computed PCollections as source and
+  prunes unnecessary pipeline parts for fast computation.
+  """
+  def __init__(
+      self,
+      user_pipeline: beam.Pipeline,
+      pcolls: Optional[Set[beam.pvalue.PCollection]] = None):
+    """
+    Initializes a pipelilne for augmenting interactive flavor.
+
+    Args:
+      user_pipeline: a beam.Pipeline instance defined by the user.
+      pcolls: cacheable pcolls to be computed/retrieved. If the set is
+        empty, all intermediate pcolls assigned to variables are applicable.
+    """
+    assert not pcolls or all([pcoll.pipeline is user_pipeline for pcoll in
+      pcolls]), 'All %s need to belong to %s' % (pcolls, user_pipeline)
+    self._user_pipeline = user_pipeline
+    self._pcolls = pcolls
+    self._cache_manager = ie.current_env().get_cache_manager(
+        self._user_pipeline, create_if_absent=True)
+    if background_caching_job.has_source_to_cache(self._user_pipeline):
+      self._cache_manager = ie.current_env().get_cache_manager(
+          self._user_pipeline)
+    self._pipeline, self._context = self._user_pipeline.to_runner_api(
+        return_context=True)
+    self._context.component_id_map = copy.copy(
+        self._user_pipeline.component_id_map)
+    self._cacheables = self.cacheables()
+    self._augmented = False
+
+  @property
+  def augmented_pipeline(self) -> beam_runner_api_pb2.Pipeline:
+    assert self._augmented, 'Call augment() before retrieving the value.'
+    return self._pipeline
+
+  # TODO(BEAM-10708): Support generating a background recording job that
+  # contains unbound source recording transforms only.
+  @property
+  def background_recording_pipeline(self) -> beam_runner_api_pb2.Pipeline:
+    raise NotImplementedError
+
+  def cacheables(self) -> Dict[beam.pvalue.PCollection, Cacheable]:
+    """Finds all the cacheable intermediate PCollections in the pipeline with
+    their metadata.
+    """
+    c = {}
+    for watching in ie.current_env().watching():
+      for key, val in watching:
+        if (isinstance(val, beam.pvalue.PCollection) and
+            val.pipeline is self._user_pipeline and
+            (not self._pcolls or val in self._pcolls)):
+          pcoll_id = self._context.pcollections.get_id(val)
+          c[val] = Cacheable(
+              pcoll_id=pcoll_id,
+              var=key,
+              pcoll=val,
+              version=str(id(val)),
+              producer_version=str(id(val.producer)))
+    return c
+
+  def augment(self) -> 'AugmentedPipeline':
+    """Augments the pipeline with cache. Idempotent and returns self.
+
+    For a cacheable PCollection, if cache exists, read cache; else, write cache.
+    """
+    if self._augmented:
+      return self
+    self._augmented = True
```

Comment:
Instead of having state that tracks whether the pipeline is augmented, it's a simpler API to have this always augment the user pipeline and let the client handle caching the result.

Reply:
That makes sense, I'll make a change to always calculate a new one.